app/com/arpnetworking/kairos/service/KairosDbServiceImpl.java

```diff
@@ -0,0 +1,256 @@
+     * Public constructor.
+     *
+     * @param kairosDbClient - Client to use to make requests to backend kairosdb
+     */
+    public KairosDbServiceImpl(final KairosDbClient kairosDbClient) {
+        this._kairosDbClient = kairosDbClient;
+    }
+
+    @Override
+    public CompletionStage<MetricsQueryResponse> queryMetricTags(final MetricsQuery query) {
+        throw new UnsupportedOperationException();
+    }
+
+    @Override
+    public CompletionStage<List<RollupTask>> queryRollups() {
+        throw new UnsupportedOperationException();
+    }
+
+    @Override
+    public CompletionStage<RollupResponse> createRollup(final RollupTask rollupTask) {
+        throw new UnsupportedOperationException();
+    }
+
+    @Override
+    public CompletionStage<RollupResponse> updateRollup(final String id, final RollupTask rollupTask) {
+        throw new UnsupportedOperationException();
+    }
+
+    @Override
+    public CompletionStage<Void> deleteRollup(final String id) {
+        throw new UnsupportedOperationException();
+    }
+
+    @Override
+    public CompletionStage<MetricsQueryResponse> queryMetrics(final MetricsQuery metricsQuery) {
+        return getMetricNames()
+                .thenApply(list -> useAvailableRollups(list, metricsQuery))
+                .thenCompose(_kairosDbClient::queryMetrics);
+    }
+
+    /**
+     * Caching metricNames call.
+     *
+     * @param containing simple string match filter for metric names
+     * @return Cached metric names, filtered by the query string.
+     */
+    @Override
+    public CompletionStage<KairosMetricNamesQueryResponse> queryMetricNames(
+            final Optional<String> containing,
+            final boolean filterRollups) {
+
+        return getMetricNames()
+                .thenApply(list -> filterMetricNames(list, containing, filterRollups))
+                .thenApply(list -> new KairosMetricNamesQueryResponse.Builder().setResults(list).build());
+    }
+
+    private static ImmutableList<String> filterMetricNames(
+            final List<String> metricNames,
+            final Optional<String> containing,
+            final boolean filterRollups) {
+
+        final Predicate<String> baseFilter;
+        if (filterRollups) {
+            baseFilter = IS_ROLLUP.negate();
+        } else {
+            baseFilter = s -> true;
+        }
+
+        final Predicate<String> containsFilter;
+
+        if (containing.isPresent() && !containing.get().isEmpty()) {
+            final String lowerContaining = containing.get().toLowerCase(Locale.ENGLISH);
+            containsFilter = s -> s.toLowerCase(Locale.getDefault()).contains(lowerContaining);
+        } else {
+            containsFilter = s -> true;
+        }
+
+        return metricNames.stream()
+                .filter(baseFilter)
+                .filter(IS_PT1M.negate())
+                .filter(containsFilter)
+                .collect(Collectors.collectingAndThen(Collectors.toList(), ImmutableList::copyOf));
```

Comment:
You should be able to do this in one step with `ImmutableList.toImmutableList()`

Reply:
I didn't use it because it was marked as @Beta but okay...